File: Data Structures & Algorithms/max-water-container/submission-6.rs
impl Solution {
    pub fn max_area(heights: Vec<i32>) -> i32 {
        let mut max_area = 0;

        let mut left = 0;
        let mut right = heights.len() - 1;

        while left < right {
            let first_bar = heights[left];
            let second_bar = heights[right];

            let curr_max_area = first_bar.min(second_bar) * (right - left) as i32;
            max_area = max_area.max(curr_max_area);

            if first_bar < second_bar {
                left += 1;
            } else {
                right -= 1
            }
        }
        return max_area;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-6.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        hash_map = {}
        longest_substring = 0
        left = 0
        max_freq = 0
        
        for right in range(len(s)):
            hash_map[s[right]] = hash_map.get(s[right], 0) + 1
            max_freq = max(max_freq, hash_map[s[right]])
            
            while (right - left + 1) - max_freq > k:
                hash_map[s[left]] -= 1
                left += 1
            
            longest_substring = max(longest_substring,  right - left + 1)
            
        return longest_substring